File: 2022/day22.cpp
// https://adventofcode.com/2022/day/22

#include "../common/common.h"
#include "../common/rust.h"

#include <algorithm>
#include <print>
#include <ranges>
#include <span>

using parsed_map_t = aoc::char_grid<>;
constexpr let empty_char = ' ';
constexpr let tile_char = '.';
constexpr let wall_char = '#';

constexpr let clockwise_char = 'R';
constexpr let counterclockwise_char = 'L';
constexpr let turn_clockwise = -1;
constexpr let turn_counterclockwise = -2;

using arrow_t = aoc::arrow_type<int>;

fn parse_map(std::span<String> raw_map, point map_size) -> parsed_map_t {
  // Add empty space all around the map to remove the need for bounds checking
  map_size += {2, 2};
  parsed_map_t parsed_map{};
  parsed_map.add_row(stdv::repeat(empty_char, map_size.x));
  for (String& line : raw_map) {
    auto empty_suffix = stdv::repeat(empty_char, map_size.x - line.size() - 1) |
                        aoc::ranges::to<String>();
    parsed_map.add_row(empty_char + std::move(line) + std::move(empty_suffix));
  }
  parsed_map.add_row(stdv::repeat(empty_char, map_size.x));
  return parsed_map;
}

struct input_t {
  parsed_map_t map;
  Vec<int> steps;
};

fn parse(String const& filename) -> input_t {
  auto steps = Vec<int>{};
  auto raw_map = Vec<String>{};

  auto parsing_steps = false;
  for (String line : aoc::views::read_lines(filename, aoc::keep_empty{},
                                            aoc::keep_spaces{})) {
    if (line.empty()) {
      parsing_steps = true;
      continue;
    }
    if (!parsing_steps) {
      raw_map.push_back(std::move(line));
    } else {
      auto start = 0;
      auto size = 0;
      let get_step_size = [&]() {
        steps.push_back(aoc::to_number<int>(line.substr(start, size)));
      };
      for (int i = 0; i < std::ssize(line); ++i) {
        let current = line[i];
        if (current == clockwise_char) {
          get_step_size();
          steps.push_back(turn_clockwise);
          start = i + 1;
          size = 0;
        } else if (current == counterclockwise_char) {
          get_step_size();
          steps.push_back(turn_counterclockwise);
          start = i + 1;
          size = 0;
        } else {
          ++size;
        }
      }
      get_step_size();
    }
  }

  let max_width =
      static_cast<int>(stdr::max_element(raw_map, {}, &String::size)->size());
  let map_size = point{max_width, static_cast<int>(raw_map.size())};
  return {parse_map(raw_map, map_size), std::move(steps)};
}

fn find_start(parsed_map_t const& parsed_map) -> arrow_t {
  let num_columns = static_cast<int>(parsed_map.num_columns());
  for (int x = 1; x < num_columns; ++x) {
    if (parsed_map.at(1, x) == tile_char) {
      return {point{x, 1}, aoc::east};
    }
  }
  return {{}, aoc::east};
}

/// Executes the step sequence:
///   - turns rotate the arrow in place
///   - move counts call step_fn(current) once per tile
///
/// step_fn returns false to stop early (hit a wall), true to continue.
template <class StepFn>
  requires requires(StepFn step_fn, arrow_t& arrow) {
    { step_fn(arrow) } -> std::convertible_to<bool>;
  }
fn execute_steps(std::span<const int> steps, arrow_t start, StepFn step_fn)
    -> arrow_t {
  auto current = start;
  for (int step : steps) {
    if (step == turn_clockwise) {
      current.direction = aoc::clockwise_basic(current.direction);
    } else if (step == turn_counterclockwise) {
      current.direction = aoc::anticlockwise_basic(current.direction);
    }
    for (int i = 0; i < step; ++i) {
      if (!step_fn(current)) {
        break;
      }
    }
  }
  return current;
}

fn password(arrow_t const& arr) -> int {
  return 1000 * arr.position.y + 4 * arr.position.x + arr.direction;
}

fn solve_case1(input_t const& input) -> int {
  let & [ parsed_map, steps ] = input;
  let start = find_start(parsed_map);

  let final_arrow = execute_steps(steps, start, [&](arrow_t& current) {
    let diff = aoc::get_diff(current.direction);
    auto next_pos = current.position + diff;
    if (parsed_map.at(next_pos.y, next_pos.x) == empty_char) {
      // Flat wrap: step back from current position to find the far edge
      auto wrapped = current.position;
      while (parsed_map.at((wrapped - diff).y, (wrapped - diff).x) !=
             empty_char) {
        wrapped = wrapped - diff;
      }
      next_pos = wrapped;
    }
    if (parsed_map.at(next_pos.y, next_pos.x) == wall_char) {
      return false;
    }
    current.position = next_pos;
    return true;
  });

  return password(final_arrow);
}

using vec3 = aoc::nd_point_type<int, 3>;
struct face_info {
  point pos;
  vec3 normal;
  vec3 right;
  vec3 down;
  constexpr bool operator==(const face_info&) const = default;
  constexpr auto operator<=>(const face_info&) const = default;
};

template <int cube_side>
fn solve_case2(input_t const& input) -> int {
  let & [ parsed_map, steps ] = input;
  let num_rows = static_cast<int>(parsed_map.num_rows());
  let num_columns = static_cast<int>(parsed_map.num_columns());

  // BFS from the first face, assigning 3D orientations to all 6 cube faces
  let start_face = [&]() -> point {
    for (int fy = 1; fy < num_rows - 1; fy += cube_side) {
      for (int fx = 1; fx < num_columns - 1; fx += cube_side) {
        if (parsed_map.at(fy, fx) != empty_char) {
          return point{fx, fy};
        }
      }
    }
    return point{-1, -1};
  }();

  // Crossing a 2D edge transforms the orientation:
  //   east:  { face.right,  -face.normal,  face.down    }
  //   west:  { -face.right,  face.normal,  face.down    }
  //   south: { face.down,    face.right,  -face.normal  }
  //   north: { -face.down,   face.right,   face.normal  }
  let faces_map = aoc::shortest_distances_dijkstra(
      face_info{
          .pos = start_face,
          .normal = {0, 0, 1},
          .right = {1, 0, 0},
          .down = {0, 1, 0},
      },
      [&](face_info const& face) {
        // face == current face
        auto neighbors = Vec<face_info>{};
        for (let dir : aoc::basic_sky_directions) {
          let diff = aoc::get_diff(dir);
          let new_pos = point{(diff.x * cube_side) + face.pos.x,
                              (diff.y * cube_side) + face.pos.y};
          if ((new_pos.x < 1) ||
              (new_pos.y < 1) ||
              (new_pos.x >= (num_columns - 1)) ||
              (new_pos.y >= (num_rows - 1))) {
            continue;
          }
          if (parsed_map.at(new_pos.y, new_pos.x) == empty_char) {
            continue;
          }
          neighbors.push_back([&] -> face_info {
            switch (dir) {
              case aoc::east:
                return {new_pos, face.right, -face.normal, face.down};
              case aoc::west:
                return {new_pos, -face.right, face.normal, face.down};
              case aoc::south:
                return {new_pos, face.down, face.right, -face.normal};
              case aoc::north:
                return {new_pos, -face.down, face.right, face.normal};
              default:
                AOC_UNREACHABLE("Invalid direction");
            }
          }());
        }
        return aoc::dijkstra_uniform_neighbors(std::move(neighbors));
      });
  auto faces = Vec<face_info>{};
  for (let& [ face, _ ] : faces_map) {
    faces.push_back(face);
  }

  // Find face by 2D position
  let face_from_pos = [&](point pos) -> const face_info& {
    return *stdr::find_if(faces, [&](let& f) {
      return (pos.x >= f.pos.x) &&
             (pos.x < (f.pos.x + cube_side)) &&
             (pos.y >= f.pos.y) &&
             (pos.y < (f.pos.y + cube_side));
    });
  };
  // Find face by 3D normal
  let face_from_normal = [&](vec3 normal) -> const face_info& {
    return *stdr::find_if(faces, [&](let& f) { return f.normal == normal; });
  };

  // Given an arrow stepping off a face edge into empty space,
  // compute the cube-wrapped destination arrow.
  // Works purely from the 3D orientations (normal/right/down)
  // assigned to each face during the BFS above.
  let cube_jump = [&](const arrow_t arrow) -> arrow_t {
    let& src = face_from_pos(arrow.position);

    // Moving east/west/south/north on the 2D map corresponds to moving in the
    // src.right / -src.right / src.down / -src.down directions in 3D.
    // The destination face is the one whose normal points that way.
    let dest_normal = [&] -> vec3 {
      switch (arrow.direction) {
        case aoc::east:
          return src.right;
        case aoc::west:
          return -src.right;
        case aoc::south:
          return src.down;
        default:
          return -src.down;
      }
    }();
    let& dest = face_from_normal(dest_normal);

    // Folding across the shared edge, the src face's outward normal becomes
    // the in-plane arrival direction on the destination face (reversed)
    let arrival_dir = -src.normal;
    auto new_facing = aoc::facing_t{};
    auto entry_coord = 0; // value of the fixed axis on dest (0 or cube_side-1)
    auto y_is_fixed = false;  // true when entering from north/south (y fixed)
    auto dest_along = vec3{}; // dest axis running parallel to the entry edge
    if (arrival_dir == dest.right) {
      new_facing = aoc::east;
      entry_coord = 0;
      y_is_fixed = false;
      dest_along = dest.down;
    } else if (arrival_dir == -dest.right) {
      new_facing = aoc::west;
      entry_coord = cube_side - 1;
      y_is_fixed = false;
      dest_along = dest.down;
    } else if (arrival_dir == dest.down) {
      new_facing = aoc::south;
      entry_coord = 0;
      y_is_fixed = true;
      dest_along = dest.right;
    } else {
      new_facing = aoc::north;
      entry_coord = cube_side - 1;
      y_is_fixed = true;
      dest_along = dest.right;
    }

    // Measure how far along the src entry edge the arrow sits,
    // then map it onto the destination edge.
    // Reverse when the two edge axes are antiparallel.
    let is_EW =
        (arrow.direction == aoc::east) || (arrow.direction == aoc::west);
    let src_along = is_EW ? src.down : src.right;
    let src_offset =
        is_EW ? (arrow.position.y - src.pos.y) : (arrow.position.x - src.pos.x);
    let dest_offset =
        (src_along == -dest_along) ? (cube_side - 1 - src_offset) : src_offset;

    let x_in_dst = y_is_fixed ? dest_offset : entry_coord;
    let y_in_dst = y_is_fixed ? entry_coord : dest_offset;
    return {dest.pos + point(x_in_dst, y_in_dst), new_facing};
  };

  let start = find_start(parsed_map);

  let final_arrow = execute_steps(steps, start, [&](arrow_t& current) {
    auto next_pos = current.position + aoc::get_diff(current.direction);
    auto next_dir = current.direction;
    if (parsed_map.at(next_pos.y, next_pos.x) == empty_char) {
      let next_arrow = cube_jump(current);
      next_pos = next_arrow.position;
      next_dir = next_arrow.direction;
    }
    if (parsed_map.at(next_pos.y, next_pos.x) == wall_char) {
      return false;
    }
    current.position = next_pos;
    current.direction = next_dir;
    return true;
  });

  return password(final_arrow);
}

int main() {
  std::println("Part 1");
  let example = parse("day22.example");
  AOC_EXPECT_RESULT(6032, solve_case1(example));
  let input = parse("day22.input");
  AOC_EXPECT_RESULT(97356, solve_case1(input));

  std::println("Part 2");
  AOC_EXPECT_RESULT(5031, (solve_case2<4>(example)));
  AOC_EXPECT_RESULT(120175, (solve_case2<50>(input)));

  AOC_RETURN_CHECK_RESULT();
}

File: rust/2022/src/day22.rs
use aoc::direction::*;
use aoc::grid::Grid;
use aoc::nd_point::NDPoint;
use aoc::point::Point;

type Pos = Point<i32>;
type Vec3 = NDPoint<i32, 3>;

const EMPTY: char = ' ';
const TILE: char = '.';
const WALL: char = '#';

const TURN_CLOCKWISE: i32 = -1;
const TURN_COUNTERCLOCKWISE: i32 = -2;

struct Input {
    map: Grid<char>,
    steps: Vec<i32>,
}

fn parse(filename: &str) -> Input {
    let mut steps = Vec::new();
    let mut raw_map: Vec<String> = Vec::new();
    let mut parsing_steps = false;

    for line in std::fs::read_to_string(filename).unwrap().lines() {
        if line.is_empty() {
            parsing_steps = true;
            continue;
        }
        if !parsing_steps {
            raw_map.push(line.to_string());
        } else {
            let mut start = 0;
            let mut size = 0;
            for (i, c) in line.char_indices() {
                match c {
                    'R' | 'L' => {
                        if size > 0 {
                            steps.push(line[start..start + size].parse().unwrap());
                        }
                        steps.push(if c == 'R' {
                            TURN_CLOCKWISE
                        } else {
                            TURN_COUNTERCLOCKWISE
                        });
                        start = i + 1;
                        size = 0;
                    }
                    _ => size += 1,
                }
            }
            if size > 0 {
                steps.push(line[start..start + size].parse().unwrap());
            }
        }
    }

    let max_width = raw_map.iter().map(|l| l.len()).max().unwrap_or(0);
    let map_width = max_width + 2;
    let map_height = raw_map.len() + 2;

    let mut map_data = Vec::with_capacity(map_width * map_height);
    map_data.extend(std::iter::repeat(EMPTY).take(map_width));
    for line in &raw_map {
        map_data.push(EMPTY);
        map_data.extend(line.chars());
        map_data.extend(std::iter::repeat(EMPTY).take(map_width - 1 - line.len()));
    }
    map_data.extend(std::iter::repeat(EMPTY).take(map_width));

    Input {
        map: Grid::from_vec(map_data, map_height, map_width),
        steps,
    }
}

fn find_start(map: &Grid<char>) -> Arrow {
    for x in 1..map.num_columns {
        if *map.get(1, x) == TILE {
            return Arrow {
                pos: Pos::new(x as i32, 1),
                dir: Direction::East,
            };
        }
    }
    panic!("no starting tile found");
}

fn execute_steps(
    steps: &[i32],
    start: Arrow,
    mut step_fn: impl FnMut(&mut Arrow) -> bool,
) -> Arrow {
    let mut current = start;
    for &step in steps {
        if step == TURN_CLOCKWISE {
            current.dir = current.dir.clockwise();
        } else if step == TURN_COUNTERCLOCKWISE {
            current.dir = current.dir.counterclockwise();
        } else {
            for _ in 0..step {
                if !step_fn(&mut current) {
                    break;
                }
            }
        }
    }
    current
}

fn password(arrow: Arrow) -> i32 {
    1000 * arrow.pos.y + 4 * arrow.pos.x + arrow.dir as i32
}

fn solve_case1(input: &Input) -> i32 {
    let Input { map, steps } = input;
    let start = find_start(map);

    let final_arrow = execute_steps(steps, start, |current| {
        let diff = current.dir.diff();
        let mut next_pos = current.pos + diff;
        if *map.get(next_pos.y as usize, next_pos.x as usize) == EMPTY {
            // Flat wrap: step back from current position to find the far edge
            let mut wrapped = current.pos;
            while *map.get((wrapped.y - diff.y) as usize, (wrapped.x - diff.x) as usize) != EMPTY {
                wrapped = wrapped - diff;
            }
            next_pos = wrapped;
        }
        if *map.get(next_pos.y as usize, next_pos.x as usize) == WALL {
            return false;
        }
        current.pos = next_pos;
        return true;
    });

    return password(final_arrow);
}

#[derive(Clone, Copy, PartialEq, Eq, Hash, PartialOrd, Ord)]
struct FaceInfo {
    pos: Pos,
    normal: Vec3,
    right: Vec3,
    down: Vec3,
}

fn new_vec3(x: i32, y: i32, z: i32) -> Vec3 {
    Vec3 { data: [x, y, z] }
}

fn solve_case2(input: &Input, cube_side: i32) -> i32 {
    let Input { map, steps } = input;
    let num_rows = map.num_rows as i32;
    let num_columns = map.num_columns as i32;

    // BFS from the first face, assigning 3D orientations to all 6 cube faces
    let start_face = {
        let mut found = None;
        'outer: for fy in (1..num_rows - 1).step_by(cube_side as usize) {
            for fx in (1..num_columns - 1).step_by(cube_side as usize) {
                if *map.get(fy as usize, fx as usize) != EMPTY {
                    found = Some(Pos::new(fx, fy));
                    break 'outer;
                }
            }
        }
        found.unwrap()
    };

    // Crossing a 2D edge transforms the orientation:
    //   east:  { face.right,  -face.normal,  face.down   }
    //   west:  { -face.right,  face.normal,  face.down   }
    //   south: { face.down,    face.right,  -face.normal }
    //   north: { -face.down,   face.right,   face.normal }
    let faces_map = aoc::dijkstra::shortest_distances(
        &FaceInfo {
            pos: start_face,
            normal: new_vec3(0, 0, 1),
            right: new_vec3(1, 0, 0),
            down: new_vec3(0, 1, 0),
        },
        |_| false,
        |face| {
            let mut neighbors = Vec::new();
            for dir in BASIC_DIRECTIONS {
                let diff = dir.diff();
                let new_pos = Pos::new(
                    (diff.x * cube_side) + face.pos.x,
                    (diff.y * cube_side) + face.pos.y,
                );
                if (false
                    || new_pos.x < 1
                    || new_pos.y < 1
                    || new_pos.x >= num_columns - 1
                    || new_pos.y >= num_rows - 1)
                {
                    continue;
                }
                if *map.get(new_pos.y as usize, new_pos.x as usize) == EMPTY {
                    continue;
                }
                let neighbor = match dir {
                    Direction::East => FaceInfo {
                        pos: new_pos,
                        normal: face.right,
                        right: -face.normal,
                        down: face.down,
                    },
                    Direction::West => FaceInfo {
                        pos: new_pos,
                        normal: -face.right,
                        right: face.normal,
                        down: face.down,
                    },
                    Direction::South => FaceInfo {
                        pos: new_pos,
                        normal: face.down,
                        right: face.right,
                        down: -face.normal,
                    },
                    Direction::North => FaceInfo {
                        pos: new_pos,
                        normal: -face.down,
                        right: face.right,
                        down: face.normal,
                    },
                };
                neighbors.push(aoc::dijkstra::DijkstraState {
                    data: neighbor,
                    distance: 1,
                });
            }
            neighbors
        },
    );

    let faces = faces_map.keys().copied().collect::<Vec<_>>();

    // Given an arrow stepping off a face edge into empty space,
    // compute the cube-wrapped destination arrow.
    // Works purely from the 3D orientations (normal/right/down)
    // assigned to each face during the BFS above.
    let cube_jump = |arrow: Arrow| -> Arrow {
        let src = *faces
            .iter()
            .find(|f| {
                return true
                    && arrow.pos.x >= f.pos.x
                    && arrow.pos.x < f.pos.x + cube_side
                    && arrow.pos.y >= f.pos.y
                    && arrow.pos.y < f.pos.y + cube_side;
            })
            .unwrap();

        // Moving east/west/south/north on the 2D map corresponds to moving in the
        // src.right / -src.right / src.down / -src.down directions in 3D.
        // The destination face is the one whose normal points that way.
        let dest_normal = match arrow.dir {
            Direction::East => src.right,
            Direction::West => -src.right,
            Direction::South => src.down,
            Direction::North => -src.down,
        };
        let dest = *faces.iter().find(|f| f.normal == dest_normal).unwrap();

        // Folding across the shared edge, the src face's outward normal becomes
        // the in-plane arrival direction on the destination face (reversed)
        let arrival_dir = -src.normal;
        let (
            new_facing, //
            entry_coord,
            y_is_fixed,
            dest_along,
        ) = if (arrival_dir == dest.right) {
            (Direction::East, 0_i32, false, dest.down)
        } else if arrival_dir == -dest.right {
            (Direction::West, cube_side - 1, false, dest.down)
        } else if arrival_dir == dest.down {
            (Direction::South, 0_i32, true, dest.right)
        } else {
            (Direction::North, cube_side - 1, true, dest.right)
        };

        // Measure how far along the src entry edge the arrow sits,
        // then map it onto the destination edge.
        // Reverse when the two edge axes are antiparallel.
        let is_ew = matches!(arrow.dir, Direction::East | Direction::West);
        let src_along = if is_ew { src.down } else { src.right };
        let src_offset = if is_ew {
            arrow.pos.y - src.pos.y
        } else {
            arrow.pos.x - src.pos.x
        };
        let dest_offset = if src_along == -dest_along {
            cube_side - 1 - src_offset
        } else {
            src_offset
        };

        let x_in_dst = if y_is_fixed { dest_offset } else { entry_coord };
        let y_in_dst = if y_is_fixed { entry_coord } else { dest_offset };
        Arrow {
            pos: dest.pos + Pos::new(x_in_dst, y_in_dst),
            dir: new_facing,
        }
    };

    let start = find_start(map);

    let final_arrow = execute_steps(steps, start, |current| {
        let next_pos = current.pos + current.dir.diff();
        let (
            final_pos, //
            final_dir,
        ) = if *map.get(next_pos.y as usize, next_pos.x as usize) == EMPTY {
            let jumped = cube_jump(*current);
            (jumped.pos, jumped.dir)
        } else {
            (next_pos, current.dir)
        };
        if *map.get(final_pos.y as usize, final_pos.x as usize) == WALL {
            return false;
        }
        current.pos = final_pos;
        current.dir = final_dir;
        return true;
    });

    return password(final_arrow);
}

fn main() {
    println!("Part 1");
    let example = parse("day22.example");
    aoc::expect_result!(6032, solve_case1(&example));
    let input = parse("day22.input");
    aoc::expect_result!(97356, solve_case1(&input));

    println!("Part 2");
    aoc::expect_result!(5031, solve_case2(&example, 4));
    aoc::expect_result!(120175, solve_case2(&input, 50));
}
